File: apps/app-frontend/src/helpers/curseforge.ts
import { invoke } from '@tauri-apps/api/core'

import type { InstallJobSnapshot } from './install'

export type ContentProvider = 'modrinth' | 'curseforge'

export interface CurseForgeCapability {
	status: 'missing_key' | 'ready' | 'unauthorized'
	configured: boolean
}

export interface CurseForgeSearchRequest {
	classId: number
	categoryId?: number
	categoryIds?: number[]
	searchFilter?: string
	gameVersion?: string
	modLoaderType?: number
	sortField?: number
	sortOrder?: 'asc' | 'desc'
	index?: number
	pageSize?: number
}

export interface UnifiedSearchHit {
	provider: 'curseforge'
	project_id: string
	slug?: string
	author: string
	author_url?: string
	title: string
	description: string
	project_type: string
	categories: string[]
	versions: string[]
	downloads: number
	icon_url?: string
	date_created: string
	date_modified: string
	latest_version?: string
	gallery: string[]
	website_url?: string
	source_url?: string
	allow_mod_distribution?: boolean
}

export interface UnifiedSearchResponse {
	provider: 'curseforge'
	hits: UnifiedSearchHit[]
	offset: number
	limit: number
	total_hits: number
}

export interface CurseForgeFilesRequest {
	gameVersion?: string
	modLoaderType?: number
	gameVersionTypeId?: number
	index?: number
	pageSize?: number
}

export interface CurseForgeProject {
	id: number
	name: string
	slug: string
	summary: string
	downloadCount: number
	mainFileId: number
	classId?: number
	dateCreated: string
	dateModified: string
	dateReleased: string
	allowModDistribution?: boolean
	gamePopularityRank?: number
	logo?: { thumbnailUrl: string; url: string }
	authors: Array<{ id: number; name: string; url: string }>
	categories: Array<{ id: number; name: string; slug: string; iconUrl?: string }>
	screenshots: Array<{ id: number; title: string; url: string; thumbnailUrl: string }>
	latestFilesIndexes: Array<{
		gameVersion: string
		fileId: number
		filename: string
		releaseType: number
		gameVersionTypeId?: number
		modLoader?: number
	}>
	links: {
		websiteUrl?: string
		wikiUrl?: string
		issuesUrl?: string
		sourceUrl?: string
	}
}

export interface CurseForgeCategory {
	id: number
	gameId: number
	name: string
	slug: string
	url: string
	iconUrl?: string
	dateModified: string
	isClass?: boolean | null
	classId?: number
	parentCategoryId?: number
	displayIndex?: number
}

export interface CurseForgeFile {
	id: number
	modId: number
	isAvailable: boolean
	displayName: string
	fileName: string
	releaseType: number
	fileDate: string
	fileLength: number
	hashes: Array<{ value: string; algo: number }>
	fileFingerprint: number
	downloadCount: number
	downloadUrl?: string
	gameVersions: string[]
	dependencies: Array<{ modId: number; relationType: number }>
}

export function getCurseForgeImageUrl(source?: string | null, width = 256): string | undefined {
	if (!source) return undefined

	try {
		const url = new URL(source)
		if (url.protocol !== 'https:' || !url.hostname.endsWith('forgecdn.net')) return source

		const proxy = new URL('https://images.weserv.nl/')
		proxy.searchParams.set('url', source)
		proxy.searchParams.set('w', String(width))
		proxy.searchParams.set('fit', 'contain')
		proxy.searchParams.set('output', 'webp')
		return proxy.toString()
	} catch {
		return source
	}
}

export interface CurseForgeFilesResponse {
	files: CurseForgeFile[]
	pagination: {
		index: number
		pageSize: number
		resultCount: number
		totalCount: number
	}
}

export interface CurseForgeInstallRequest {
	instanceId: string
	projectId: number
	fileId: number
	projectType: string
	ownershipKind?: 'pack_managed' | 'user_added'
	manualOperationKind?: 'pack_install' | 'pack_update' | 'content_install' | 'content_update'
	gameVersion?: string
	modLoaderType?: number
	worldName?: string
	installDependencies?: boolean
}

export interface CurseForgeInstallResult {
	installed: Array<{
		projectId: number
		fileId: number
		relativePath: string
		dependency: boolean
	}>
	manualDownloads: Array<{
		projectId: number
		fileId: number
		fileName: string
		ownershipKind: 'pack_managed' | 'user_added'
		operationKind: 'pack_install' | 'pack_update' | 'content_install' | 'content_update'
		websiteUrl?: string
		projectType: string
		projectSlug: string
		targetFolder: string
		hashes: Array<{ value: string; algo: number }>
		fileLength: number
		fileFingerprint: number
	}>
	failedDownloads: Array<{
		projectId: number
		fileId: number
		fileName: string
		reason: string
	}>
	optionalDependencies: number[]
	incompatibleDependencies: number[]
}

export interface CurseForgeManualDownloadImport {
	projectId: number
	fileId: number
	relativePath: string
}

export interface CurseForgeManualDownloadScanResult {
	downloadDirectory?: string | null
	imported: CurseForgeManualDownloadImport[]
	errors: Array<{
		projectId: number
		fileId: number
		message: string
	}>
}

export interface CurseForgeModpackInstallResult {
	content: CurseForgeInstallResult
	overridesWritten: number
	minecraftVersion: string
	loader?: string
}

export function summarizeCurseForgeInstall(result: CurseForgeInstallResult) {
	const installed = result.installed?.length ?? 0
	const manual = result.manualDownloads?.length ?? 0
	const failed = result.failedDownloads?.length ?? 0
	const optional = result.optionalDependencies?.length ?? 0
	const incompatible = result.incompatibleDependencies?.length ?? 0
	return { installed, manual, failed, optional, incompatible }
}

export function getCurseForgeDownloadFailureDetails(error: unknown): string | null {
	const message =
		error instanceof Error
			? error.message
			: typeof error === 'string'
				? error
				: typeof error === 'object' && error !== null && 'message' in error
					? String(error.message)
					: null
	if (!message) return null

	const normalized = message.toLowerCase()
	const isDownloadFailure = normalized.includes('download failed after')
	const isCurseForge =
		normalized.includes('curseforge') ||
		normalized.includes('forgecdn.net') ||
		normalized.includes('forgecdn')

	return isDownloadFailure && isCurseForge ? message : null
}

export function getCurseForgeCapability() {
	return invoke<CurseForgeCapability>('plugin:curseforge|curseforge_capability')
}

export function validateCurseForgeCredentials() {
	return invoke<CurseForgeCapability>('plugin:curseforge|curseforge_validate_credentials')
}

export function searchCurseForgeProjects(request: CurseForgeSearchRequest, requestId?: string) {
	return invoke<UnifiedSearchResponse>('plugin:curseforge|curseforge_search_projects', {
		request,
		requestId,
	})
}

export function getCurseForgeProject(projectId: number) {
	return invoke<CurseForgeProject>('plugin:curseforge|curseforge_get_project', { projectId })
}

export function getCurseForgeDescription(projectId: number) {
	return invoke<string>('plugin:curseforge|curseforge_get_description', {
		projectId,
	})
}

export function getCurseForgeFiles(projectId: number, request: CurseForgeFilesRequest) {
	return invoke<CurseForgeFilesResponse>('plugin:curseforge|curseforge_get_files', {
		projectId,
		request,
	})
}

export function getCurseForgeFile(projectId: number, fileId: number) {
	return invoke<CurseForgeFile>('plugin:curseforge|curseforge_get_file', {
		projectId,
		fileId,
	})
}

export function getCurseForgeDownloadUrl(projectId: number, fileId: number) {
	return invoke<string | null>('plugin:curseforge|curseforge_get_download_url', {
		projectId,
		fileId,
	})
}

export function getCurseForgeCategories(classId?: number) {
	return invoke<CurseForgeCategory[]>('plugin:curseforge|curseforge_get_categories', { classId })
}

export function installCurseForgeFile(request: CurseForgeInstallRequest) {
	return invoke<CurseForgeInstallResult>('plugin:curseforge|curseforge_install_file', { request })
}

export function queueCurseForgeFile(
	request: CurseForgeInstallRequest,
	display: { title: string; iconUrl?: string | null },
) {
	return invoke<InstallJobSnapshot>('plugin:instance|instance_queue_curseforge_content', {
		request,
		displayTitle: display.title,
		displayIcon: display.iconUrl ?? null,
	})
}

export function updateCurseForgeFile(instanceId: string, relativePath: string) {
	return invoke<CurseForgeInstallResult>('plugin:curseforge|curseforge_update_installed_file', {
		instanceId,
		relativePath,
	})
}

export function switchCurseForgeFileVersion(
	instanceId: string,
	relativePath: string,
	fileId: number,
) {
	return invoke<CurseForgeInstallResult>(
		'plugin:curseforge|curseforge_switch_installed_file_version',
		{
			instanceId,
			relativePath,
			fileId,
		},
	)
}

export function recognizeCurseForgeFiles(instanceId: string) {
	return invoke<{
		scanned: number
		matched: number
		linked: CurseForgeInstallResult['installed']
		unmatchedPaths: string[]
	}>('plugin:curseforge|curseforge_recognize_instance_files', { instanceId })
}

export function importCurseForgeManualDownloads(
	instanceId: string,
	downloads: CurseForgeInstallResult['manualDownloads'],
) {
	return invoke<CurseForgeManualDownloadScanResult>(
		'plugin:curseforge|curseforge_import_manual_downloads',
		{ instanceId, downloads },
	)
}

export function installCurseForgeModpack(request: {
	instanceId: string
	projectId: number
	fileId: number
	installOptional?: boolean
}) {
	return invoke<CurseForgeModpackInstallResult>('plugin:curseforge|curseforge_install_modpack', {
		request,
	})
}

export function updateManagedCurseForgeModpack(instanceId: string, fileId: number) {
	return invoke<CurseForgeModpackInstallResult>(
		'plugin:curseforge|curseforge_update_managed_modpack',
		{
			instanceId,
			fileId,
		},
	)
}

File: apps/app/src/api/curseforge.rs
use crate::api::search_cancellation::SearchCancellation;
use crate::api::{Result, TheseusSerializableError};
use theseus::curseforge::{
    CurseForgeCapability, CurseForgeCategory, CurseForgeFile,
    CurseForgeFilesRequest, CurseForgeFilesResponse,
    CurseForgeFingerprintResult, CurseForgeInstallRequest,
    CurseForgeInstallResult, CurseForgeManualDownload,
    CurseForgeManualDownloadScanResult, CurseForgeModpackInstallRequest,
    CurseForgeModpackInstallResult, CurseForgeProject,
    CurseForgeRecognitionResult, CurseForgeSearchRequest,
    UnifiedSearchResponse,
};

pub fn init<R: tauri::Runtime>() -> tauri::plugin::TauriPlugin<R> {
    tauri::plugin::Builder::new("curseforge")
        .invoke_handler(tauri::generate_handler![
            curseforge_capability,
            curseforge_validate_credentials,
            curseforge_search_projects,
            curseforge_get_project,
            curseforge_get_projects,
            curseforge_get_description,
            curseforge_get_files,
            curseforge_get_file,
            curseforge_get_files_many,
            curseforge_get_changelog,
            curseforge_get_download_url,
            curseforge_get_categories,
            curseforge_match_fingerprints,
            curseforge_install_file,
            curseforge_update_installed_file,
            curseforge_switch_installed_file_version,
            curseforge_recognize_instance_files,
            curseforge_import_manual_downloads,
            curseforge_install_modpack,
            curseforge_update_managed_modpack,
        ])
        .build()
}

#[tauri::command]
pub fn curseforge_capability() -> CurseForgeCapability {
    theseus::curseforge::capability()
}

#[tauri::command]
pub async fn curseforge_validate_credentials() -> Result<CurseForgeCapability> {
    Ok(theseus::curseforge::validate_credentials().await?)
}

#[tauri::command]
pub async fn curseforge_search_projects(
    request: CurseForgeSearchRequest,
    request_id: Option<String>,
) -> Result<UnifiedSearchResponse> {
    let Some(request_id) = request_id else {
        return Ok(theseus::curseforge::search_projects(request).await?);
    };
    let cancellation = SearchCancellation::register(request_id);

    tokio::select! {
        result = theseus::curseforge::search_projects(request) => Ok(result?),
        _ = cancellation.cancelled() => Err(TheseusSerializableError::SearchCancelled(
            "CurseForge browse search".to_string(),
        )),
    }
}

#[tauri::command]
pub async fn curseforge_get_project(
    project_id: u32,
) -> Result<CurseForgeProject> {
    Ok(theseus::curseforge::get_project(project_id).await?)
}

#[tauri::command]
pub async fn curseforge_get_projects(
    project_ids: Vec<u32>,
) -> Result<Vec<CurseForgeProject>> {
    Ok(theseus::curseforge::get_projects(project_ids).await?)
}

#[tauri::command]
pub async fn curseforge_get_description(project_id: u32) -> Result<String> {
    Ok(theseus::curseforge::get_description(project_id).await?)
}

#[tauri::command]
pub async fn curseforge_get_files(
    project_id: u32,
    request: CurseForgeFilesRequest,
) -> Result<CurseForgeFilesResponse> {
    Ok(theseus::curseforge::get_files(project_id, request).await?)
}

#[tauri::command]
pub async fn curseforge_get_file(
    project_id: u32,
    file_id: u32,
) -> Result<CurseForgeFile> {
    Ok(theseus::curseforge::get_file(project_id, file_id).await?)
}

#[tauri::command]
pub async fn curseforge_get_files_many(
    file_ids: Vec<u32>,
) -> Result<Vec<CurseForgeFile>> {
    Ok(theseus::curseforge::get_files_many(file_ids).await?)
}

#[tauri::command]
pub async fn curseforge_get_changelog(
    project_id: u32,
    file_id: u32,
) -> Result<String> {
    Ok(theseus::curseforge::get_changelog(project_id, file_id).await?)
}

#[tauri::command]
pub async fn curseforge_get_download_url(
    project_id: u32,
    file_id: u32,
) -> Result<Option<String>> {
    Ok(theseus::curseforge::get_download_url(project_id, file_id).await?)
}

#[tauri::command]
pub async fn curseforge_get_categories(
    class_id: Option<u32>,
) -> Result<Vec<CurseForgeCategory>> {
    Ok(theseus::curseforge::get_categories(class_id).await?)
}

#[tauri::command]
pub async fn curseforge_match_fingerprints(
    fingerprints: Vec<u64>,
) -> Result<CurseForgeFingerprintResult> {
    Ok(theseus::curseforge::match_fingerprints(fingerprints).await?)
}

#[tauri::command]
pub async fn curseforge_install_file(
    request: CurseForgeInstallRequest,
) -> Result<CurseForgeInstallResult> {
    Ok(theseus::curseforge::install_file(request).await?)
}

#[tauri::command]
pub async fn curseforge_update_installed_file(
    instance_id: String,
    relative_path: String,
) -> Result<CurseForgeInstallResult> {
    Ok(
        theseus::curseforge::update_installed_file(
            &instance_id,
            &relative_path,
        )
        .await?,
    )
}

#[tauri::command]
pub async fn curseforge_switch_installed_file_version(
    instance_id: String,
    relative_path: String,
    file_id: u32,
) -> Result<CurseForgeInstallResult> {
    Ok(theseus::curseforge::switch_installed_file_version(
        &instance_id,
        &relative_path,
        file_id,
    )
    .await?)
}

#[tauri::command]
pub async fn curseforge_recognize_instance_files(
    instance_id: String,
) -> Result<CurseForgeRecognitionResult> {
    Ok(theseus::curseforge::recognize_instance_files(&instance_id).await?)
}

#[tauri::command]
pub async fn curseforge_import_manual_downloads(
    instance_id: String,
    downloads: Vec<CurseForgeManualDownload>,
) -> Result<CurseForgeManualDownloadScanResult> {
    Ok(
        theseus::curseforge::import_manual_downloads(&instance_id, downloads)
            .await?,
    )
}

#[tauri::command]
pub async fn curseforge_install_modpack(
    request: CurseForgeModpackInstallRequest,
) -> Result<CurseForgeModpackInstallResult> {
    Ok(theseus::curseforge::install_modpack(request).await?)
}

#[tauri::command]
pub async fn curseforge_update_managed_modpack(
    instance_id: String,
    file_id: u32,
) -> Result<CurseForgeModpackInstallResult> {
    Ok(
        theseus::curseforge::update_managed_modpack(&instance_id, file_id)
            .await?,
    )
}
